### ISSUE TITLE
[flang][OpenMP] Missing diagnostic for bare DECLARE TARGET in invalid program units

### ISSUE THREAD

**Issue by MattPD** [llvm/llvm-project#198038]
The bare form of `!$omp declare target` (no arguments, no clauses) is only permitted in the specification part of a subroutine, function, or interface body (OpenMP 5.2, section 7.8.2).

Flang currently accepts it without error in BLOCK DATA, PROGRAM, MODULE, SUBMODULE, and BLOCK constructs.

**Expected behavior:** Flang should emit a diagnostic when the bare form appears outside a valid scope (subroutine, function, or interface body).

**Reproducer:**

```fortran
block data my_block_data
  implicit none
  integer :: x
  !$omp declare target  ! Should be rejected
end block data

program main
  implicit none
  !$omp declare target  ! Should be rejected
end program

module my_mod
  implicit none
  !$omp declare target  ! Should be rejected
end module
```

**Cross-compiler behavior:**

- gfortran trunk, rejects: https://godbolt.org/z/Ef81nvecs
- Intel ifx, rejects: https://godbolt.org/z/3qKjo1ecb
- nvfortran 26.3, accepts: https://godbolt.org/z/9dYK64nTq
- LLVM Flang trunk, accepts: https://godbolt.org/z/acE4vPoKx
- CCE (ftn): rejects with `ftn-1064: The OpenMP DECLARE TARGET (without clauses) directive must not be specified inside MODULE/SUBMODULE/PROGRAM`

Only nvfortran and (old) Flang accept the bare form in these scopes. The fix should align Flang with the majority compiler behavior (gfortran, ifx, CCE) and the OpenMP specification.
